pootle/static/js/browser/components/BrowserTable.js

```diff
@@ -0,0 +1,328 @@
+                <tbody>
+                    <tr>
+                        <td className="translated" style={{'width': trans_p+'%'}}><span></span></td>
+                        <td className="fuzzy" style={{'width': fuzzy_p+'%'}}><span></span></td>
+                        <td className="untranslated"><span></span></td>
+                    </tr>
+                </tbody>
+            </table>
+        );
+    }
+});
+
+const BrowserTable = React.createClass({
+
+    propTypes: {
+        items: React.PropTypes.array.isRequired,
+    },
+
+    // TODO: is there a better way to define constants (not fields)?
+    COL_TITLE: 0,
+    COL_PROGRESS: 1,
+    COL_TOTAL: 2,
+    COL_LASTUPDATED: 3,
+    COL_CRITICAL: 4,
+    COL_SUGGESTIONS: 5,
+    COL_INCOMPLETE: 6,
+    COL_LASTACTIVITY: 7,
+
+    getInitialState : function() {
+        return {
+            showDisabledRows: false,
+            sortColumn: this.COL_TITLE,
+            reverseSort: false,
+        };
+    },
+
+    componentWillMount() {
+        // check if there are disabled items
+        this.hasDisabledItems = false;
+        for (var key in this.props.items) {
+            var item = this.props.items[key];
+            if (item.is_disabled) {
+                this.hasDisabledItems = true;
+                break;
+            }
+        }
+
+        this.sortedKeys = Object.keys(this.props.items).sort(this.sortFunc);
+    },
+
+    handleDisabledRowsVisibility({ isActive: state }) {
+        var s = this.state;
+        s.showDisabledRows = state;
+        this.setState(s);
+    },
+
+    sortFunc(a, b) {
+        var ia = this.props.items[a];
+        var ib = this.props.items[b];
+
+        if (this.state.sortColumn == this.COL_TITLE) {
+            return ia.title.localeCompare(ib.title);
+        }
+
+        if (this.state.sortColumn == this.COL_PROGRESS) {
+            return ia.progress - ib.progress;
+        }
+
+        if (this.state.sortColumn == this.COL_TOTAL) {
+            return ia.total - ib.total;
+        }
+
+        if (this.state.sortColumn == this.COL_LASTUPDATED) {
+            return ib.lastupdated - ia.lastupdated; // reversed
+        }
+
+        if (this.state.sortColumn == this.COL_CRITICAL) {
+            return ia.critical - ib.critical;
+        }
+
+        if (this.state.sortColumn == this.COL_SUGGESTIONS) {
+            return ia.suggestions - ib.suggestions;
+        }
+
+        if (this.state.sortColumn == this.COL_INCOMPLETE) {
+            return ia.incomplete - ib.incomplete;
+        }
+
+        if (this.state.sortColumn == this.COL_LASTACTIVITY) {
+            return ib.lastaction.mtime - ia.lastaction.mtime; // reversed
+        }
+
+        throw new Error('Sort column index out of range: '+this.state.sortColumn);
+    },
+
+    handleHeaderClick(colN) {
+        var s = this.state;
+        if (s.sortColumn == colN) {
+            // quick reverse
+            s.reverseSort = !s.reverseSort;
```

Comment:
Let's never ever mutate state directly. We can make a copy, and then `setState()`.